### PR TITLE
Features/chapter4

### PR DESCRIPTION
**Pull request by voldem007**

### CODE REVIEW

File: SimpleRss/SimpleRss/Views/TopicViewCell.swift [hunk 1]
```diff
@@ -11,4 +11,29 @@ import UIKit
...
+    weak var getOperation: GetImageOperation?
+    
+    var imageUrl: URL? {
+        didSet {
+            previewImageView.image = nil
+            
+            getOperation?.cancel()
+            downloadOperation?.cancel()
+            
+            let getOp = GetImageOperation((imageUrl?.lastPathComponent)!)
```
Comment (warokku) [inline SimpleRss/SimpleRss/Views/TopicViewCell.swift]: Please remove force unwrapping here

File: SimpleRss/SimpleRss/Views/TopicViewCell.swift [hunk 2]
```diff
@@ -11,4 +11,29 @@ import UIKit
...
+            getOperation = getOp
+            
+            let downloadOp = DownloadImageOperation(imageUrl!)
```
Comment (warokku) [inline SimpleRss/SimpleRss/Views/TopicViewCell.swift]: And here

File: SimpleRss/SimpleRss/Views/FeedViewCell.swift [hunk 3]
```diff
@@ -15,10 +15,19 @@ class FeedViewCell: UITableViewCell {
...
     @IBOutlet weak var pubDateLabel: UILabel!
     
+    var getOperation: GetImageOperation?
+    
     var imageUrl: URL? {
         didSet {
-            previewImageView.image = nil
-            updateImage()
+            guard let url = imageUrl else { return }
+            ImageDownloadOrchestrator.shared.cancel(getOperation)
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/Views/FeedViewCell.swift]: Cancel operation in prepareForReuse. Same for TopicViewCell

File: SimpleRss/SimpleRss/Services/ImageDownloadOrchestrator.swift [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    let cacheManager: CacheManager
+    let downloadManager: DownloadManager
+    
+    required init(cacheManager: CacheManager, downloadManager: DownloadManager) {
+        self.cacheManager = cacheManager
+        self.downloadManager = downloadManager
+    }
+    
+    func download(url: URL, _ completion: @escaping(_ image: UIImage?) -> Void) -> GetImageOperation {
+        let getOp = cacheManager.get(url: url)
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/Services/ImageDownloadOrchestrator.swift:25]: Why not to check if file exists on the disk before that?

File: SimpleRss/SimpleRss/Views/Operations/DownloadImageOperation.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        self.url = url
+        super.init()
+    }
+    
+    override func start() {
+        super.start()
+        
+        let image = imageLoad(imageUrl: url)
+        
+        if isCancelled {
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/Views/Operations/DownloadImageOperation.swift]: This chunk of code would make sense if you were using NSURLSession and could cancel dowloading operation. Same for GetImageOperation

File: SimpleRss/SimpleRss/Services/ImageDownloadOrchestrator.swift [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        getOp.completionBlock = { [weak self] in
+            guard let self = self else { return }
+            if !getOp.isCancelled {
```
Comment (Viktor-by) [inline SimpleRss/SimpleRss/Services/ImageDownloadOrchestrator.swift]: Move that check to `guard let`

**Review by Viktor-by — CHANGES_REQUESTED** [review]

File: SimpleRss/SimpleRss/Services/ImageDownloadOrchestrator.swift (hunk 4 above)
Comment (voldem007) [reply SimpleRss/SimpleRss/Services/ImageDownloadOrchestrator.swift:25]: insert check for exist into getOperation

**Review by warokku — APPROVED** [review]

**Review by Viktor-by — APPROVED** [review]

File: SimpleRss/SimpleRss/Views/TopicViewCell.swift (hunk 2 above)
Comment (voldem007) [reply SimpleRss/SimpleRss/Views/TopicViewCell.swift]: looks like previous changes, won't fix

File: SimpleRss/SimpleRss/Views/TopicViewCell.swift (hunk 1 above)
Comment (voldem007) [reply SimpleRss/SimpleRss/Views/TopicViewCell.swift]: looks like previous changes, won't fix

**Pull request merged**